js/src/data/actions.js

```diff
@@ -72,77 +73,80 @@ export function* fetchShippingRates() {
 }
 
 /**
- * Aggregated shipping rate.
+ * Saves shipping rates.
  *
- * @typedef {Object} AggregatedShippingRate
- * @property {Array<CountryCode>} countries Array of destination country codes.
- * @property {string} currency Currency of the price.
- * @property {number} rate Shipping price.
- */
-
-/**
- * Updates or inserts given aggregated shipping rate.
+ * This is done by removing the old shipping rates first,
+ * and then upserting the new shipping rates.
  *
- * @param {AggregatedShippingRate} shippingRate
+ * @param {Array<ShippingRate>} newShippingRates
  */
-export function* upsertShippingRates( shippingRate ) {
-	const { countryCodes, currency, rate } = shippingRate;
-
+export function* saveShippingRates( newShippingRates ) {
 	try {
-		yield apiFetch( {
-			path: `${ API_NAMESPACE }/mc/shipping/rates/batch`,
-			method: 'POST',
-			data: {
-				country_codes: countryCodes,
-				currency,
-				rate,
-			},
-		} );
+		/**
+		 * Compare the new shipping rates and the old ones from the store
+		 * to find out the old ones to be deleted.
+		 */
+		const oldShippingRates = yield select( STORE_KEY, 'getShippingRates' );
+		const deleteIds = oldShippingRates.reduce( ( acc, cur ) => {
+			const found = newShippingRates.find( ( el ) =>
+				isSameShippingRate( el, cur )
+			);
+
+			if ( ! found ) {
+				acc.push( cur.id );
+			}
 
-		return {
-			type: TYPES.UPSERT_SHIPPING_RATES,
-			shippingRate,
-		};
-	} catch ( error ) {
-		yield handleFetchError(
-			error,
-			__(
-				'There was an error trying to add / update shipping rates. Please try again later.',
-				'google-listings-and-ads'
-			)
-		);
-	}
-}
+			return acc;
+		}, [] );
+
+		if ( deleteIds.length ) {
+			yield apiFetch( {
+				path: `${ API_NAMESPACE }/mc/shipping/rates/batch`,
+				method: 'DELETE',
+				data: {
+					ids: deleteIds,
+				},
+			} );
+		}
 
-/**
- * Deletes shipping rates associated with given country codes.
- *
- * @param {Array<CountryCode>} countryCodes
- */
-export function* deleteShippingRates( countryCodes ) {
-	try {
-		yield apiFetch( {
+		/**
+		 * No new shipping rate, no need to call API.
+		 */
+		if ( ! newShippingRates.length ) {
+			return {
+				type: TYPES.RECEIVE_SHIPPING_RATES,
+				shippingRates: [],
+			};
+		}
+
+		/**
+		 * There are new shipping rates, we call API to upsert them.
+		 */
+		const data = yield apiFetch( {
 			path: `${ API_NAMESPACE }/mc/shipping/rates/batch`,
-			method: 'DELETE',
+			method: 'POST',
 			data: {
-				country_codes: countryCodes,
+				rates: newShippingRates,
 			},
 		} );
 
+		const upsertedShippingRates = data.success?.map( ( el ) => el.rate );
```

Comment:
❓ or 🚧  Is there any case the API will respond without `success` property? If so, `data.success?.map()` will result in `undefined` and break the page.

Reply:
Good thinking! I looked into the PHP code (in `feature/refactor-shipping-rates-api` branch) and it seems like `success` array should always be there, so we should be safe here. 😄 

https://github.com/woocommerce/google-listings-and-ads/blob/4bfc1c723b7a80414b40bb578178a765ebaeed7b/src/API/Site/Controllers/MerchantCenter/ShippingRateBatchController.php#L48-L74

Reply:
Gotcha, done in [`b01c91b` (#1262)](https://github.com/woocommerce/google-listings-and-ads/pull/1262/commits/b01c91ba31d64c36c403b30a41e830b29ff17d72). 🙂